### PR TITLE
JBTM-3398 Test LRA behaviour on different Object Stores

### PR DESCRIPTION
**Pull request by jmfinelli**
This PR resolves the JIRA issue [JBTM-3398](https://issues.redhat.com/browse/JBTM-3398)

LRA JDK11
!MAIN !CORE !TOMCAT !AS_TESTS !RTS !JACOCO !XTS !QA_JTA !QA_JTS_JACORB !QA_JTS_JDKORB !QA_JTS_OPENJDKORB !BLACKTIE !PERF LRA NO_WIN !DB_TESTS

### CODE REVIEW

**Comment by jmfinelli** [comment]
TESTIT

**Comment by ochaloup** [comment]
TESTIT

**Comment by mmusgrov** [comment]
>  narayana CI not started.
> 
> Author is not the 'narayana' contributor, to permit PR being run members of jbosstm can write comment of text: TESTIT

Strange, I will take a look.

**Comment by jmfinelli** [comment]
TESTIT

**Comment by jmfinelli** [comment]
TESTIT

**Comment by jmfinelli** [comment]
TESTIT

**Review by ochaloup — CHANGES_REQUESTED** [review]
I have few questions and remarks. The main thing which I think needs to be changed is the way how the test is executed. There need to be a possibility to run with filesystem objectore (as default) and as second option with JDBC.
From what I can see the PR changes the default option for JDBC and does not provide a way to run with file system, does it?

File: rts/lra/coordinator/src/test/java/io/narayana/lra/coordinator/LRACoordinatorRecovery2TestCase.java
Comment (jmfinelli) [reply rts/lra/coordinator/src/test/java/io/narayana/lra/coordinator/LRACoordinatorRecovery2TestCase.java:89]: Thanks, I will tune the timeout ASAP.

This delay was introduced because the test was randomly failing to recover the Long & Short transactions. I am still investigating why this happened.

File: rts/lra/coordinator/src/test/resources/scripts/participant-byteman-rules.btm
Comment (jmfinelli) [reply rts/lra/coordinator/src/test/resources/scripts/participant-byteman-rules.btm:9]: same problem of https://github.com/jbosstm/narayana/pull/1765#discussion_r561788476

File: rts/lra/coordinator/pom.xml
Comment (jmfinelli) [reply rts/lra/coordinator/pom.xml:71]: [this instruction](https://github.com/jmfinelli/narayana/blob/4a2742d3d6288f8440fd8ade0fe8ae4eb086490a/rts/lra/coordinator/src/test/java/io/narayana/lra/coordinator/JDBCTestBaseImpl.java#L58) was not working without this dependency

File: rts/lra/coordinator/src/test/java/io/narayana/lra/coordinator/AbstractTestBase.java
Comment (jmfinelli) [reply rts/lra/coordinator/src/test/java/io/narayana/lra/coordinator/AbstractTestBase.java:249]: I'll change that :+1:

File: rts/lra/coordinator/src/test/java/io/narayana/lra/coordinator/FileSystemTestBaseImpl.java
Comment (jmfinelli) [reply rts/lra/coordinator/src/test/java/io/narayana/lra/coordinator/FileSystemTestBaseImpl.java:41]: No problem :+1:

File: rts/lra/coordinator/src/test/java/io/narayana/lra/coordinator/FileSystemTestBaseImpl.java
Comment (jmfinelli) [reply rts/lra/coordinator/src/test/java/io/narayana/lra/coordinator/FileSystemTestBaseImpl.java:61]: I agree :+1:

**Comment by jmfinelli** [comment]
> [...] There need to be a possibility to run with filesystem objectore (as default) and as second option with JDBC.
> From what I can see the PR changes the default option for JDBC and does not provide a way to run with file system, does it?

Yes, you are right. There is an implementation of the class `AbstractTestBase` to run tests using the File System but it needs to be chosen at compile time.

> Maybe there could be two surefire executions defined or a special profile which will switch from test to jdbc.

This sounds like a nice solution

File: rts/lra/coordinator/pom.xml
Comment (ochaloup) [reply rts/lra/coordinator/pom.xml:71]: ok,  I see. Thanks for elaboration. Maybe just the scope could be "downgraded" to `test`?

Comment (jmfinelli) [reply rts/lra/coordinator/pom.xml:71]: that makes sense :+1: thanks

**Comment by mmusgrov** [comment]
@jmfinelli Normally we wait for the CI tests to pass before submitting a review, unless the engineer is reasonably confident that it will pass. But it is also fine to ask for an early review if the intent is to solicit feedback as the fix progresses.

File: rts/lra/coordinator/src/test/resources/standalone-default.xml
Comment (jmfinelli) [reply rts/lra/coordinator/src/test/resources/standalone-default.xml:1]: I agree with you @ochaloup. The reason why the CI failed was an older version of the XML configuration file (I was using Wildfly 21). As soon as I updated the XML file, the test passed.

@mmusgrov could give us your inputs on this?

**Comment by mmusgrov** [comment]
I have added the Hold label to this PR pending the reworking of the test to use Undertow instead of WildFly as discussed elsewhere.
@jmfinelli Please remove the Hold label when you are ready for another round of reviews, thanks.

**Comment by jmfinelli** [comment]
I have a solution (nearly) ready but I am waiting for [your PR](https://github.com/jbosstm/narayana/pull/1764) to be merged as I based my solution on your PR. I'll keep the HOLD label until your PR gets approved.

**Review by mmusgrov — CHANGES_REQUESTED** [review]
You don't need to change all of these classes. You just a single test that looks similar to the following (with a couple of pom includes for Undertow):

```java
        UndertowJaxrsServer server = new UndertowJaxrsServer().start();
        server.deploy(Coordinator.class);
        BeanPopulator.getDefaultInstance(ObjectStoreEnvironmentBean.class).setObjectStoreType(jdbcstore);
        BeanPopulator.getDefaultInstance(ObjectStoreEnvironmentBean.class).configure tables etc
        // take a look at some of our existing JDBCStore unit tests
        URI lraid = lraClient.startLRA(...);
        lraClient.getStatus(lraid) or run an sql statement against the JDBC store
```

**Comment by mmusgrov** [comment]
My PR (#1764 ) with the undertow dependencies has addressed all of the reviewers' comments so I'd expect it to be merged today. You can either wait for that to be merged and pull from master (git pull --rebase upstream master) or you can just add the undertow dependency to your pom:

``` java
        <dependency>
            <groupId>org.jboss.resteasy</groupId>
            <artifactId>resteasy-undertow</artifactId>
            <version>${version.org.jboss.resteasy}</version>
            <scope>test</scope>
        </dependency>
```

**Comment by jmfinelli** [comment]
> My PR (#1764 ) with the undertow dependencies has addressed all of the reviewers' comments so I'd expect it to be merged today. You can either wait for that to be merged and pull from master (git pull --rebase upstream master) or you can just add the undertow dependency to your pom:
> 
> ```java
>         <dependency>
>             <groupId>org.jboss.resteasy</groupId>
>             <artifactId>resteasy-undertow</artifactId>
>             <version>${version.org.jboss.resteasy}</version>
>             <scope>test</scope>
>         </dependency>
> ```

If we can wait, I would delay pushing my solution until your PR gets approved

File: rts/lra/coordinator/src/test/java/io/narayana/lra/coordinator/domain/model/LRATest.java
Comment (mmusgrov) [inline rts/lra/coordinator/src/test/java/io/narayana/lra/coordinator/domain/model/LRATest.java:821]: Would it also be possible to run an SQL query against the JDBC object store to verify that the record made it to the database as expected.

Maybe also try closing the LRA and then use another SQL query to validate that the transaction log has been removed.

Maybe also check that the expected attributes of a LongRunningAction are populated in the result set of an SQL query.

File: rts/lra/coordinator/src/test/java/io/narayana/lra/coordinator/domain/model/LRATest.java
Comment (mmusgrov) [inline rts/lra/coordinator/src/test/java/io/narayana/lra/coordinator/domain/model/LRATest.java:828]: Why did you choose to test the file system store but none of the others. I would have expected a test against just the JDBC store or all of the supported ones - on wildfly we support the JDBC, filesystem and journal stores. But, yes, I like the idea of sanity checking all of the supported stores.

**Review by mmusgrov — CHANGES_REQUESTED** [review]

File: rts/lra/coordinator/src/test/java/io/narayana/lra/coordinator/domain/model/LRATest.java
Comment (jmfinelli) [reply rts/lra/coordinator/src/test/java/io/narayana/lra/coordinator/domain/model/LRATest.java:821]: Thanks for your review :-) No problem to add the SQL queries.

Just to double check, I thought that once I am sure that the Object Store is set to the desired type (in this case, JDBCStore), the method getStatus(URI uri) of NarayanaLRAClient already verifies the existence of the record for us. So, for example, if I close the LRA and then I try to get its status again, I should get a NotFoundException.

File: rts/lra/coordinator/src/test/java/io/narayana/lra/coordinator/domain/model/LRATest.java
Comment (jmfinelli) [reply rts/lra/coordinator/src/test/java/io/narayana/lra/coordinator/domain/model/LRATest.java:828]: Yeah, it is true, I should have checked other object store types. I'll update the PR :-)

File: rts/lra/coordinator/src/test/java/io/narayana/lra/coordinator/domain/model/LRATest.java
Comment (mmusgrov) [reply rts/lra/coordinator/src/test/java/io/narayana/lra/coordinator/domain/model/LRATest.java:821]: Yes, that's the theory but the goal of testing is to verify the hypotheses that the developer got it right. Examples:

- some strange execution path ends up writing the log in a different store type;
- some of the attributes of the record are missing or have the wrong value; 
- the record gets written to the wrong table;
- etch

Comment (jmfinelli) [reply rts/lra/coordinator/src/test/java/io/narayana/lra/coordinator/domain/model/LRATest.java:821]: Thanks for your comment @mmusgrov, I'll add some code to programmatically check our hypotheses :+1:

File: rts/lra/coordinator/src/test/java/io/narayana/lra/coordinator/domain/model/objectstore/TestBase.java
Comment (mmusgrov) [inline rts/lra/coordinator/src/test/java/io/narayana/lra/coordinator/domain/model/objectstore/TestBase.java:107]: Remark: the list must contain at least one since a) the LRA has not closed (would be removed from the store), b) has not failed (would be moved to another part of the store) and c) has not timed out (would have been removed from the store). In order words your implicit assumption that the LRA still exists is valid (that said if any of those assumptions were invalidated then you would see a messy IndexOutOfBoundsException in the test output).

**Review by mmusgrov — APPROVED** [review]
Thanks.
@ochaloup I think this one is good to merge now.

File: rts/lra/coordinator/src/test/java/io/narayana/lra/coordinator/domain/model/objectstore/TestBase.java
Comment (mmusgrov) [reply rts/lra/coordinator/src/test/java/io/narayana/lra/coordinator/domain/model/objectstore/TestBase.java:107]: Ah wait, you are testing if `LRAList.isEmpty()` so you would never get an IndexOutOfBoundsException.
Apologies for my invalid remark.

Comment (jmfinelli) [reply rts/lra/coordinator/src/test/java/io/narayana/lra/coordinator/domain/model/objectstore/TestBase.java:107]: Yep, I made sure I was not fetching anything out of boundaries

**Review by ochaloup — APPROVED** [review]
I would only be afraid of fact the tests do not clean the state to time when it was started (e.g. the jdbc storage seems to be used when test finishes). But it seems it does not make any issue so we can live with it.

**Comment by ochaloup** [comment]
we've discussed my point about the cleaning with @jmfinelli and I understood that the surefire forks the JVM for each class and thus the setup done in the `@BeforeClass` is vital only for particular class and does not influence the other tests

**Comment by jmfinelli** [comment]
> We've discussed my point about the cleaning with @jmfinelli and I understood that the surefire forks the JVM for each class and thus the setup done in the `@BeforeClass` is vital only for particular class and does not influence the other tests. My point about cleaning is probably not valid then.

Summarising the discussion @ochaloup and I had:
the surefire plugin in the root pom is configured to fork a JVM for each test class through `<forkMode>pertest</forkMode>`

**Pull request merged**
